### DISCUSSION TITLE
Format of the tomolist.txt file?

### DISCUSSION THREAD

**Discussion by JuhaHuiskonen** [nextpyp/.github#73] [Q&A]
@abartesaghi could you please provide a description of the `original_volumes.txt` file format? Specifically, what is the convention of the `normalX`, `normalY` and `normalZ` fields? Are these used to determine the orientation of the particle in Euler angles?  If yes, what is the convention? Is it the same as in Frealign for instance?  Are values in degrees or radians? Can the matrix fields used alternatively? What is the order of elements (which 3x3 values specify the rotation and which 3 values specify the translation)?  This information would help others to write particle metadata conversion tools from other packages. Many thanks!

**Comment by tofushan** [comment]
Yes, you can convert your angles into the convention used in the txt. However, converting the translations is quite tricky, because this file was generated by one of our old programs for sub-volume averaging, and its convention is not intuitive. 

If you only have rotations, you can parse the three Euler angles (in degrees) in normX/Y/Z:
normX: rotation around X axis 
normY/Z: rotation around Z axis

They all follow left handedness (clockwise positive). If we apply the rotation to **the reference**, the order would be -normY, -normX, -normZ. 

However, if you also have translation, we do not recommend doing so. We suggest you to convert them to RELION first and then convert them back to nextPYP.

**Comment by JuhaHuiskonen** [comment]
I have only rotations in RELION format (Rot, Tilt, Psi) which are applied around Z, Y and then (new) Z axes (ZYZ convention).  Do I understand correctly from your description that the angles in columns normZ, normX, normY are applied around Z, X and then (new) Z axes (ZXZ convention)?

**Reply by tofushan** [reply]
Yes, you are right. You can convert them into ZXZ convention (normY, X, Z in order).

**Comment by JuhaHuiskonen** [comment]
Just to make sure I am getting it right, could you please look into this package https://github.com/alisterburt/eulerangles/blob/master/eulerangles/constants.py and let me know if this additional code I am planning to use would match your conventions:

```
'nextpyp': ConversionMeta(name='nextpyp',
                             axes='zxz',
                             intrinsic=False,
                             right_handed_rotation=False,
                             active=False)

```

Also, when I convert the input angles (rot, tilt, psi in ZYZ convention) to lets say alpha, beta, gamma (in ZXZ convention) do I need to write these as beta = normalX,  alpha=normalY, gamma=normalZ (because the order is different in nextPYP)? Thanks a lot, almost there I hope!

**Reply by tofushan** [reply]
I think it is correct, but you may need to try different combinations like swapping normalX and Z because I'm not sure which (alpha or gamma) comes first. 

However, since you have a RELION star file, you can convert them back to a nextPYP project.

**Reply by JuhaHuiskonen** [reply]
I can see the option Import -> Tomography (from star). But how could I use this in practise in my case?  I don't want to import the entire project from RELION (I don't have it, I haven't used RELION to process these data - I have picked membranes in Dynamo and converted this metadata to STAR format). Next I just want to add the orientation information to my current project in nextPYP, using a STAR file that has `rot, tilt, psi`.

**Reply by tofushan** [reply]
I see. You can try the following combinations, at least one will work: 

1. alpha=normalY, beta=normalX, gamma=normalZ
2. alpha=normalZ, beta=normalX, gamma=normalY
3. alpha=-normalY, beta=-normalX, gamma=-normalZ
4. alpha=-normalZ, beta=-normalX, gamma=-normalY

**Reply by JuhaHuiskonen** [reply]
Still no luck.  Is it possible to use the `matrix` fields to give a rotation matrix, instead of the three angles?

**Reply by tofushan** [reply]
Yes, you can do that, but you still need to be sure about the order. 

The matrix is made up of the elements in the txt:
m[0], m[1], m[2], m[3]
m[4], m[5], m[6], m[7]
m[8], m[9], m[10], m[11]
m[12], m[13], m[14], m[15]

Could you elaborate on how you convert rotation from convention ZYZ to ZXZ? Maybe a snapshot of your code would be helpful.

**Reply by JuhaHuiskonen** [reply]
I am using the `eulerangles` package I mentioned above and input angles directly from Dynamo (ZXZ convention) or after converting my angles to RELION (ZYZ convention).  This package can convert correctly between these two. The difference between Dynamo and nextPYP seems to be that netxPYP has left-handed rotations whereas they are right-handed in Dynamo. Also the order of angles in normalX, normalY and normalZ fields is not standard (the second rotation to be applied is in `normalX` column). I would like to check one thing: in nextPYP `normalX`, `normalY`, `normalZ` are the angles supposed to be in degrees (as I have been assuming) or in radians?

My current attempt for a conversion script is here:  [dynamo2nextpyp.py](https://github.com/LSB-Helsinki/dynamo2nextpyp)

**Reply by tofushan** [reply]
The normal angles in nextPYP are all in degrees, and `normalX`, `normalY` and `normalZ` are `X`, `Z1`, `Z2` respectively. 

If the original form of rotation you have is from Dynamo, which is also ZXZ convention, you don't need the package to convert angles. Trying the four combinations suggested in my previous response and this response should work, unless Dynamo adopts additional rotations/translations to the sub-volumes. Lets say your Dynamo angles are `tdrot`, `tilt` and `narot` in ZXZ:

1. `normalY` = `tdrot`; `normalX` = `tilt`; `normalZ` = `narot`
2. `normalY` = `narot`; `normalX` = `tilt`; `normalZ` = `tdrot`
3. `normalY` = `-tdrot`; `normalX` = `-tilt`; `normalZ` = `-narot`
4. `normalY` = `-narot`; `normalX` = `-tilt`; `normalZ` = `-tdrot`

Note that there are minus signs in the third and forth combinations, which are meant to try different handedness.
